### PR TITLE
Loadcaffe

### PR DESCRIPTION
**Pull request by jenniew**
## What changes were proposed in this pull request?
Add python load caffe feature, load torch/bigdl model, also some fundamental layers.

## How was this patch tested?
add unit test.

### CODE REVIEW

File: dl/src/main/python/nn/layer.py [hunk 1]
```diff
@@ -337,6 +392,14 @@ def __init__(self,
                                    num_input_dims)
 
 
+def build_model():
```
Comment (jason-dai) [inline dl/src/main/python/nn/layer.py]: why do we have this method here?

File: dl/src/main/python/nn/layer.py [hunk 2]
```diff
@@ -82,9 +98,14 @@ def test(self, val_rdd, batch_size, val_methods):
...
-    def from_path(path, bigdl_type="float"):
-        jmodel = callBigDlFunc(bigdl_type, "modelFromPath", path)
-        return Model.of(jmodel, bigdl_type)
+    def from_path(path, model_type="bigdl", bigdl_type="float"):
+        jmodel = callBigDlFunc(bigdl_type, "modelFromPath", path, model_type)
+        return Model.of(jmodel)
+
+    @staticmethod
+    def load_caffe(model, defPath, modelPath, match_all=True, bigdl_type="float"):
+        jmodel = callBigDlFunc(bigdl_type, "loadCaffe", model, defPath, modelPath, match_all)
```
Comment (jason-dai) [inline dl/src/main/python/nn/layer.py]: model.value?

File: dl/src/main/python/nn/layer.py (hunk 1 above)
Comment (jenniew) [inline dl/src/main/python/nn/layer.py]: This is a utility function for loadcaffe doctest, as we already have separate loadcaffe test, I remove it.

File: dl/src/main/python/nn/layer.py [hunk 3]
```diff
@@ -82,9 +82,14 @@ def test(self, val_rdd, batch_size, val_methods):
                              val_rdd, batch_size, val_methods)
 
     @staticmethod
-    def from_path(path, bigdl_type="float"):
-        jmodel = callBigDlFunc(bigdl_type, "modelFromPath", path)
-        return Model.of(jmodel, bigdl_type)
+    def from_path(path, model_type="bigdl", bigdl_type="float"):
```
Comment (jason-dai) [inline dl/src/main/python/nn/layer.py]: change to:
1)load
2)load_torch
3)local_caffe
@zhichao-li

Comment (jenniew) [inline dl/src/main/python/nn/layer.py]: updated.

File: dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala [hunk 4]
```diff
@@ -970,10 +970,14 @@ class PythonBigDL[T: ClassTag](implicit ev: TensorNumeric[T]) extends Serializab
     testResultArray.toList.asJava
   }
 
-  def modelFromPath(path: String): AbstractModule[Activity, Activity, T] = {
+  def loadBigdl(path: String): AbstractModule[Activity, Activity, T] = {
```
Comment (jason-dai) [inline dl/src/main/scala/com/intel/analytics/bigdl/python/api/PythonBigDL.scala]: loadBigDL

File: dl/src/main/python/nn/layer.py [hunk 5]
```diff
@@ -415,11 +424,9 @@ class View(Model):
 
     def __init__(self,
                  sizes,
-                 num_input_dims=0,
```
Comment (zhichao-li) [inline dl/src/main/python/nn/layer.py]: let's keep this default parameters.

File: dl/src/main/python/nn/layer.py [hunk 6]
```diff
@@ -415,11 +424,9 @@ class View(Model):
...
                  bigdl_type="float"):
         super(View, self).__init__(None, bigdl_type,
-                                   sizes,
-                                   num_input_dims)
```
Comment (zhichao-li) [inline dl/src/main/python/nn/layer.py]: keep this.

**Comment by zhichao-li** [comment]
@jenniew let's rebase and merge it after pass the jenkins.

**Comment by zhichao-li** [comment]
Squash those commits into one before merging it.

**Pull request merged**